rfd/0001-runjob-design.md

```diff
@@ -0,0 +1,360 @@
+- command - A child process of `runjob`
+
+This will be the approximate process for executing a Job in the `jobmanager` server:
+
+- The library will call the `jobmanager` executable binary with an alternate mode called `runjob`, using `os/exec.Command()`. It will pass the Job command and command arguments as an argument.
+- The library in the parent `jobmanager` process will also set `SysProcAttr.Pdeathsig = syscall.SIGKILL` on its `runjob` child process.  If the parent server process dies, the kernel will kill the child process and all of it's decendents as the parent process will have orphaned processes.
+- As part of the `exec.Command()`, the library in the `runjob` child process will add process attributes via `syscall.SysProcAttr`for creating new namespaces for PID, mount, and network for the underlying `runjob` process being created.
+- The library in the `runjob` child process will also use `syscall.SysProcAttr` attributes of `Setpgid: true` to create a new process group for the child process. This will allow the library to more easily manage any processes that this child process creates.
+- The library in the `runjob` child process will also set `SysProcAttr.Pdeathsig = syscall.SIGQUIT` so that if the parent process dies without cleaning-up any of its child processes, the kernel will send a `SIGQUIT` signal to the child processes so that they can exit gracefully.
+- `runjob` will duplicate `STDERR` to a new file descriptor using `FD_CLOEXEC` to close it once the Job command is executed.  This sets up the Job command to capture `STDOUT` and `STDERR` to the same stream, which is then streamed back to the parent process.
+- `runjob` will utilize the library to mount a new proc filesystem, and then fork and execute the command and its arguments from the Job
+- If the process is launched successfully, the library will return an `io.ReadCloser` which can be read for the combined `STDOUT` and `STDERR` stream.
+
+There might be some other related details to work through while implementing this functionality, but generally speaking, this will be the process for executing jobs.
+
+### Job Output Streaming
+
+The library will stream the combined output of `STDOUT` and `STDERR` of running processes in an `io.ReadCloser` back to the server. The library and server will be connected to the same pipe created before the `jobmanager runjob` is called to start the requested process.
+
+The server will use a pub/sub model to distribute Jobs' output to clients. Each Job's output will be stored in its own associated buffer, which can then be replayed to one or many Subscribers (clients). The Jobs' buffer will support concurrent reads and writes via a mutex.  Subscribers can be created at any time and will be fed all of the contents in the buffer created since the start of the Job.
```

Comment:
> The Jobs' buffer will support concurrent reads and writes via a mutex.
> Subscribers can be created at any time and will be fed all of the contents in the buffer created since the start of the Job.

How will you ensure that all clients are able to read the output independently, i.e without waiting on each other?
And how will you ensure that readers can't block writing new command output?

For example, suppose there is a long-running job and there are several subscribers that have already consumed all of the command output buffer. Then a new subscriber is added. This new subscriber is quite slow though (maybe their connection isn't great). While the new subscriber "catches up" reading the output buffer, more output is generated.
How can we ensure that the slow reader doesn't prevent the others from seeing the latest output?

Reply:
Ah right, good call.  I will add more detail here around the buffer having only one reader and one writer doing locks. That way, a mechanism above that buffer is responsible for the fan out process.  Thereby not creating concurrent blocking reads of the buffer.

Reply:
I have been thinking about this more and working on some code in this area and see exactly what you mean WRT slow readers causing deadlocks.  After working through code that essentially copies out chunks of a `bytes.Buffer` and used a `sync.RWMutex`, it just never seemed to work well and always had readers deadlocking.

Moved it to using a `[]byte` instead as a buffer, using offsets and chunks, and it seems to be working well.   Will update the doc accordingly.